Source/DafnyLanguageServer/DafnyLanguageServer.cs

```diff
@@ -19,21 +20,45 @@ public static class DafnyLanguageServer {
       }
     }
 
-    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options, IConfiguration configuration) {
+    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options,
+      IConfiguration configuration, CancellationTokenSource cancelLanguageServer) {
       return options
         .WithDafnyLanguage(configuration)
         .WithDafnyWorkspace(configuration)
         .WithDafnyHandlers()
-        .OnInitialize(InitializeAsync)
+        .OnInitialize((server, @params, token) => InitializeAsync(server, @params, token, cancelLanguageServer))
         .OnStarted(StartedAsync);
     }
 
-    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancellationToken) {
+    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancelRequestToken,
+        CancellationTokenSource cancelLanguageServer) {
       var logger = server.GetRequiredService<ILogger<Program>>();
       logger.LogTrace("initializing service");
+
+      KillLanguageServerIfParentDies(request, cancelLanguageServer);
```

Comment:
I'm very new to this code, so I need to better understand it.
Why was the request not used before? Is this code run for every request?

Reply:
The request is also used in another location, but this method allows you do additional work. An `InitializeParams` request is sent at most once, so this code is run at most once.